File: app/src/main/java/com/capstone/bankit/ui/expense/ExpenseViewModel.java
package com.capstone.bankit.ui.expense;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.capstone.bankit.data.api.ApiClient;
import com.capstone.bankit.ui.expense.data.ApiService;
import com.capstone.bankit.ui.expense.data.ReceiptResponse;

import java.io.File;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ExpenseViewModel extends ViewModel {
    public MutableLiveData<File> selectedImageFile = new MutableLiveData<>();
    public MutableLiveData<String> uploadResult = new MutableLiveData<>();

    public void uploadReceipt(File file) {
        ApiService receiptService = ApiClient.getInstance().create(ApiService.class);

        // Convert file to MultipartBody.Part
        RequestBody requestFile = RequestBody.create(MediaType.parse("image/jpeg"), file);
        MultipartBody.Part body = MultipartBody.Part.createFormData("receipt", file.getName(), requestFile);

        // Call API
        receiptService.uploadReceipt(body).enqueue(new Callback<ReceiptResponse>() {
            @Override
            public void onResponse(Call<ReceiptResponse> call, Response<ReceiptResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    uploadResult.postValue(response.body().getData().getUrl());
                } else {
                    uploadResult.postValue("Failed: " + response.message());
                }
            }

            @Override
            public void onFailure(Call<ReceiptResponse> call, Throwable t) {
                uploadResult.postValue("Error: " + t.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/capstone/bankit/ui/expense/ExpenseViewModel.kt
package com.capstone.bankit.ui.expense

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.capstone.bankit.data.models.ExpenseDetailResponse
import com.capstone.bankit.data.models.ExpensePostResponse
import com.capstone.bankit.data.models.ExpenseRequest
import com.capstone.bankit.data.models.ReceiptCheckResponse
import com.capstone.bankit.data.repository.BankitRepository
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.asRequestBody
import java.io.File
import java.io.IOException

class ExpenseViewModel(private val bankitRepository: BankitRepository) : ViewModel() {
    var selectedImageFile: MutableLiveData<File> = MutableLiveData()

    private val _receiptUploadUrl = MutableLiveData<String>()
    val receiptUploadUrl: LiveData<String> = _receiptUploadUrl

    private val _receiptUploadPath = MutableLiveData<String>()
    val receiptUploadPath: LiveData<String> = _receiptUploadPath

    private val _expenseDetail = MutableLiveData<ExpenseDetailResponse>()
    val expenseDetail: LiveData<ExpenseDetailResponse> = _expenseDetail

    fun postReceipt(token: String, receiptFile: File, onFailure: (String) -> Unit) {
        viewModelScope.launch {
            try {
                val requestBody = receiptFile.asRequestBody("image/jpeg".toMediaTypeOrNull())
                val receiptPart = MultipartBody.Part.createFormData("receipt", receiptFile.name, requestBody)

                val response = bankitRepository.postReceipt(token, receiptPart)

                _receiptUploadUrl.postValue(response.data?.url.toString())
                _receiptUploadPath.postValue(response.data?.filename.toString())
            } catch (e: IOException) {
                onFailure(e.message.toString())
            } catch (e: Exception) {
                onFailure(e.message.toString())
            }
        }
    }

    fun postExpense(
        token: String,
        request: ExpenseRequest,
        onSuccess: (ExpensePostResponse) -> Unit,
        onFailure: (String) -> Unit
    ){
        viewModelScope.launch {
            try {
                val response = bankitRepository.postExpense(token, request)
                onSuccess(response)
            } catch (e: IOException) {
                onFailure(e.message.toString())
            } catch (e: Exception) {
                onFailure(e.message.toString())
            }
        }
    }

    fun getExpenseDetail(
        token: String,
        expenseId: String,
        onSuccess: () -> Unit,
        onFailure: (String) -> Unit
    ){
        viewModelScope.launch {
            try {
                val response = bankitRepository.getExpenseDetail(token, expenseId)
                _expenseDetail.value = response
                onSuccess()
            } catch (e: IOException) {
                onFailure(e.message.toString())
            } catch (e: Exception) {
                onFailure(e.message.toString())
            }
        }
    }

    fun getReceipt(
        token: String,
        filename: String,
        onSuccess: (ReceiptCheckResponse) -> Unit,
        onFailure: (String) -> Unit
    ){
        viewModelScope.launch {
            try {
                val response = bankitRepository.receiptCheck(token, filename)
                onSuccess(response)
            } catch (e: IOException) {
                onFailure(e.message.toString())
            } catch (e: Exception) {
                onFailure(e.message.toString())
            }
        }
    }

}